In [1]:
import pandas as pd
import os

In [2]:
#school_data_to_load = "Resources/schools_complete.csv"
#student_data_to_load = "Resources/students_complete.csv"
school_data_to_load = os.path.join("Resources", "schools_complete.csv")
student_data_to_load = os.path.join("Resources", "students_complete.csv")

In [3]:
school_data_df = pd.read_csv(school_data_to_load)
student_data_df = pd.read_csv(student_data_to_load)

In [4]:
school_data_df.count()

School ID      15
school_name    15
type           15
size           15
budget         15
dtype: int64

In [5]:
school_data_df.isnull().sum()

School ID      0
school_name    0
type           0
size           0
budget         0
dtype: int64

In [6]:
school_data_df.notnull().sum()

School ID      15
school_name    15
type           15
size           15
budget         15
dtype: int64

In [7]:
school_data_df.dtypes

School ID       int64
school_name    object
type           object
size            int64
budget          int64
dtype: object

In [8]:
prefixes_suffixes = ["Dr. ", "Mr. ","Ms. ", "Mrs. ", "Miss ", " MD", " DDS", " DVM", " PhD"]

for word in prefixes_suffixes:
    student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")
    
student_data_df.head()

/var/folders/cp/fx9y5wj56w38_vpnhqbp7zj40000gp/T/ipykernel_6455/2445393736.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")


Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  
0          79  
1          61  
2          60  
3          58  
4          84

In [9]:
school_data_complete_df = pd.merge(student_data_df, school_data_df, on=["school_name"])
school_data_complete_df.head()

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
3          58          0  District  2917  1910635  
4          84          0  District  2917  1910635

In [10]:
student_count = school_data_complete_df["Student ID"].count()
print(f"Total number of students is {student_count:,}")

Total number of students is 39,170


In [11]:
school_count = school_data_df["school_name"].count()
print(f"Totoal number of schools is {school_count:,}")

Totoal number of schools is 15


In [12]:
school_count2 = school_data_complete_df["school_name"].unique()
school_count2

array(['Huang High School', 'Figueroa High School', 'Shelton High School',
       'Hernandez High School', 'Griffin High School',
       'Wilson High School', 'Cabrera High School', 'Bailey High School',
       'Holden High School', 'Pena High School', 'Wright High School',
       'Rodriguez High School', 'Johnson High School', 'Ford High School',
       'Thomas High School'], dtype=object)

In [13]:
total_budget = school_data_df["budget"].sum()
print(f"Total budget is {total_budget}")

Total budget is 24649428


In [14]:
average_reading_score = school_data_complete_df["reading_score"].mean()
print(f"Average reading score is {average_reading_score:.2f}")

Average reading score is 81.88


In [15]:
average_math_score = school_data_complete_df["math_score"].mean()
print(f"Average math score is {average_math_score:.2f}")

Average math score is 78.99


In [16]:
pass_math_bool = school_data_complete_df["math_score"] >= 70

pass_math = school_data_complete_df[pass_math_bool]
pass_math.head()

Student ID   student_name gender grade        school_name  reading_score  \
0           0   Paul Bradley      M   9th  Huang High School             66   
4           4     Bonnie Ray      F   9th  Huang High School             97   
5           5  Bryan Miranda      M   9th  Huang High School             94   
6           6  Sheena Carter      F  11th  Huang High School             82   
8           8   Michael Roth      M  10th  Huang High School             95   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635

In [17]:
pass_reading_bool = school_data_complete_df["reading_score"] >= 70

pass_reading = school_data_complete_df[pass_reading_bool]
pass_reading.head()

Student ID     student_name gender grade        school_name  reading_score  \
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   

   math_score  School ID      type  size   budget  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635

In [18]:
pass_math_counts = pass_math["student_name"].count()
pass_reading_counts = pass_reading["student_name"].count()

print(f"Total number of passing math: {pass_math_counts:,}")
print(f"Total number of passing reading: {pass_reading_counts:,}")

Total number of passing math: 29,370
Total number of passing reading: 33,610


In [19]:
pass_math_percentage = pass_math_counts / float(student_count) * 100
pass_reading_percentage = pass_reading_counts / float(student_count) * 100

print(f"Percentage of passing math: {pass_math_percentage:.2f}%")
print(f"Percentage of passing reading: {pass_reading_percentage:.2f}%")

Percentage of passing math: 74.98%
Percentage of passing reading: 85.81%


In [20]:
overall_pass_math_reading = school_data_complete_df[pass_reading_bool & pass_math_bool]
overall_pass_math_reading.head()

Student ID    student_name gender grade        school_name  reading_score  \
4           4      Bonnie Ray      F   9th  Huang High School             97   
5           5   Bryan Miranda      M   9th  Huang High School             94   
6           6   Sheena Carter      F  11th  Huang High School             82   
8           8    Michael Roth      M  10th  Huang High School             95   
9           9  Matthew Greene      M  10th  Huang High School             96   

   math_score  School ID      type  size   budget  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635  
9          84          0  District  2917  1910635

In [21]:
overall_pass_math_reading_counts = overall_pass_math_reading["student_name"].count()

print(f"Total number of passing both math and reading: {overall_pass_math_reading_counts:,}")

Total number of passing both math and reading: 25,528


In [22]:
overall_pass_math_reading_percentage = overall_pass_math_reading_counts / float(student_count) * 100

print(f"Percentage of passing both math and readig is {overall_pass_math_reading_percentage:.2f}%")

Percentage of passing both math and readig is 65.17%


In [23]:
district_summary_df = pd.DataFrame(
          [{"Total Schools": school_count,
          "Total Students": student_count,
          "Total Budget": total_budget,
          "Average Math Score": average_math_score,
          "Average Reading Score": average_reading_score,
          "% Passing Math": pass_math_percentage,
         "% Passing Reading": pass_reading_percentage,
        "% Overall Passing": overall_pass_math_reading_percentage}])
district_summary_df

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

my_grades = [92.34, 84.56, 86.78, 98.32]

for grade in my_grades:
    print("{:.0f}".format(grade))

my_grades = ['B', 'C', 'B' , 'D']
my_grades = pd.Series(my_grades)
my_grades.map({'B': 'A', 'C': 'B', 'D': 'C'})

In [24]:
district_summary_df.dtypes

Total Schools              int64
Total Students             int64
Total Budget               int64
Average Math Score       float64
Average Reading Score    float64
% Passing Math           float64
% Passing Reading        float64
% Overall Passing        float64
dtype: object

district_summary_df['Total Students'] = pd.to_numeric(district_summary_df["Total Students"], errors='coerce')
district_summary_df["Total Students"] = district_summary_df["Total Students"].fillna(39170)
district_summary_df["Total Students"]

In [25]:
district_summary_df["Total Students"] = district_summary_df["Total Students"].map("{:,.0f}".format)
district_summary_df["Total Students"]

0    39,170
Name: Total Students, dtype: object

In [26]:
district_summary_df["Total Budget"] = district_summary_df["Total Budget"].map("${:,.2f}".format)
district_summary_df["Total Budget"]

0    $24,649,428.00
Name: Total Budget, dtype: object

In [27]:
district_summary_df["Average Math Score"] = district_summary_df["Average Math Score"].map("{:.1f}".format)

district_summary_df["Average Reading Score"] = district_summary_df["Average Reading Score"].map("{:.1f}".format)

district_summary_df["% Passing Math"] = district_summary_df["% Passing Math"].map("{:.0f}".format)

district_summary_df["% Passing Reading"] = district_summary_df["% Passing Reading"].map("{:.0f}".format)

district_summary_df["% Overall Passing"] = district_summary_df["% Overall Passing"].map("{:.0f}".format)

In [28]:
district_summary_df

Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00               79.0   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                  81.9             75                86                65

In [29]:
district_summary_df.dtypes

Total Schools             int64
Total Students           object
Total Budget             object
Average Math Score       object
Average Reading Score    object
% Passing Math           object
% Passing Reading        object
% Overall Passing        object
dtype: object

In [30]:
new_column_order = ["Total Schools", "Total Students", "Total Budget","Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

district_summary_df = district_summary_df[new_column_order]
district_summary_df

Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00               79.0   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                  81.9             75                86                65